examples/shader/shader_defs.rs

```diff
@@ -84,94 +55,44 @@ fn setup(mut commands: Commands, mut meshes: ResMut<Assets<Mesh>>) {
     });
 }
 
-struct IsRedPipeline {
-    mesh_pipeline: MeshPipeline,
-    shader: Handle<Shader>,
-}
-
-impl FromWorld for IsRedPipeline {
-    fn from_world(world: &mut World) -> Self {
-        let asset_server = world.resource::<AssetServer>();
-        let mesh_pipeline = world.resource::<MeshPipeline>();
-        let shader = asset_server.load("shaders/shader_defs.wgsl");
-        IsRedPipeline {
-            mesh_pipeline: mesh_pipeline.clone(),
-            shader,
-        }
+impl Material for CustomMaterial {
+    fn fragment_shader() -> ShaderRef {
+        "shaders/shader_defs.wgsl".into()
     }
-}
-
-impl SpecializedMeshPipeline for IsRedPipeline {
-    type Key = (IsRed, MeshPipelineKey);
 
     fn specialize(
-        &self,
-        (is_red, pbr_pipeline_key): Self::Key,
-        layout: &MeshVertexBufferLayout,
-    ) -> Result<RenderPipelineDescriptor, SpecializedMeshPipelineError> {
-        let mut shader_defs = Vec::new();
-        if is_red.0 {
-            shader_defs.push("IS_RED".to_string());
+        _pipeline: &MaterialPipeline<Self>,
+        descriptor: &mut RenderPipelineDescriptor,
+        _layout: &MeshVertexBufferLayout,
+        key: MaterialPipelineKey<Self>,
+    ) -> Result<(), SpecializedMeshPipelineError> {
+        if key.bind_group_data.is_red {
+            let fragment = descriptor.fragment.as_mut().unwrap();
+            fragment.shader_defs.push("IS_RED".to_string());
         }
-        let mut descriptor = self.mesh_pipeline.specialize(pbr_pipeline_key, layout)?;
-        descriptor.vertex.shader = self.shader.clone();
-        descriptor.vertex.shader_defs = shader_defs.clone();
-        let fragment = descriptor.fragment.as_mut().unwrap();
-        fragment.shader = self.shader.clone();
-        fragment.shader_defs = shader_defs;
-        descriptor.layout = Some(vec![
-            self.mesh_pipeline.view_layout.clone(),
-            self.mesh_pipeline.mesh_layout.clone(),
-        ]);
-        Ok(descriptor)
+        Ok(())
     }
 }
 
-type DrawIsRed = (
-    SetItemPipeline,
-    SetMeshViewBindGroup<0>,
-    SetMeshBindGroup<1>,
-    DrawMesh,
-);
+// This is the struct that will be passed to your shader
+#[derive(AsBindGroup, TypeUuid, Debug, Clone)]
+#[uuid = "f690fdae-d598-45ab-8225-97e2a3f056e0"]
+#[bind_group_data(CustomMaterialKey)]
```

Comment:
What is the `CustomMaterialKey` used for in this example?

My initial guess (due to the name `Key` and the `Hash` impl) was that it's used as some form of change detection to determine if the binding should be updated.  I then found the documentation for the `AsBindGroup` trait, which makes it seem like `CustomMaterialKey` is just extra data, but I still don't understand how it's useful here.

Thanks for the help!

Reply:
Added some comments to explain the purpose of the key here.